File: v1/src/main/java/com/api/v1/utils/RentFinderUtil.java
package com.api.v1.utils;

import com.api.v1.domain.Rent;
import com.api.v1.domain.RentRepository;
import com.api.v1.exceptions.RentNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.math.BigInteger;

@Component
public class RentFinderUtil {

    @Autowired
    private RentRepository repository;

    public Mono<Rent> find(String orderNumber) {
        return repository.
                findAll()
                .filter(e -> e.getOrderNumber().equals(new BigInteger(orderNumber)))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new RentNotFoundException(orderNumber)));
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/rent/utils/RentFinderUtil.kt
package com.api.v2.rent.utils

import com.api.v2.rent.exceptions.RentNotFoundException
import com.api.v2.rent.domain.Rent
import com.api.v2.rent.domain.RentRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.singleOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.math.BigInteger

@Component
class RentFinderUtil {

    @Autowired
    private lateinit var repository: RentRepository

    suspend fun find(orderNumber: String): Rent {
        return withContext(Dispatchers.IO) {
            val existingRent = repository
                .findAll()
                .filter { e -> e.orderNumber == BigInteger(orderNumber) }
                .singleOrNull()
            if (existingRent == null) {
                throw RentNotFoundException(orderNumber)
            }
            existingRent
        }
    }

}